plutus-book/doc/auction/statemachine.adoc

```diff
@@ -0,0 +1,72 @@
+The contract must somehow keep track of bids, because we need to make sure that
+only bidders who are _not_ currently the highest bidder can reclaim their bids,
+that the highest bidder can claim the token in the end and so on.
+
+This means the contract needs _state_, which is more complicated than just the
+value being locked into it. In the last chapter, we could work with trivial `()`
+data scripts, because all the state was encoded in the locked value. In this
+chapter, we really need non-trivial state in our data scripts. This state will
+change through the lifetime of the auction, depending on which actions (bid,
+reclaim etc.) are performed.
+
+Abstractly, a _state machine_ is given by a set of possible _states_, an
+_initial state_ and - for each state - as set of possible _inputs_ (or
+_actions_). Given a state and an input for that state, the _transition function_
+determines the next state.
+
+Luckily, there is support for state machine contracts in
+module `Language.PlutusTx.StateMachine`:
+
+First of all, a _state machine_ is defined as
+
+[source, haskell]
+----
+data StateMachine s i = StateMachine 
+    { smTransition :: s -> i -> Maybe s
+    , smCheck      :: s -> i -> PendingTx -> Bool
+    }
+----
```

Comment:
Hm, this sort of thing is a bit risky, since we have no automated way of ensuring that this continues to match the actual definition. I don't have a good suggestion, though. Some less good suggestions:
- Turn `StateMachine.hs` into a literate module and include it into the book.
- Just don't include copies of definitions?

Reply:
The first suggestion might be worth considering. The second one won't work - of course we need to include some definitions. Like `WalletAPI` in the off-chain chapter and similar things.